### PR TITLE
Update: Include debugging information when rule throws error

### PR DESCRIPTION
**Pull request by pfhayes**
**What is the purpose of this pull request? (put an "X" next to item)**

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[x] Add something to the core
[ ] Other, please explain:

**What changes did you make? (Give an overview)**

When developing eslint rules (including for plugins), tracking down
errors within rules is difficult and requires manual updating of the
code to track down where the error is coming from. This adds additional
diagnostic information which can help

```
  eslint:traverser An error occurred while traversing +736ms
  eslint:traverser Filename: /Users/patrick/dev/test.js +0ms
  eslint:traverser Parser Options: { ecmaFeatures: { globalReturn: false, jsx: true },
  codeFrame: false,
  ecmaVersion: 6,
  sourceType: 'module' } +0ms
  eslint:traverser Parser Path: /Users/patrick/dev/node_modules/babel-eslint/index.js +2ms
  eslint:traverser Settings: { react: { pragma: 'React' } } +0ms
```

**Is there anything you'd like reviewers to focus on?**

I opted for `debug` logging to make this minimally disruptive, but I
would also advocate for promoting this information to being always logged to
stderr. This information will be valuable to developers working on
rules, but may also be valuable to the end developer to help them track
down the source of their error

### CODE REVIEW

**Comment by platinumazure** [comment]
**TSC Summary:** There is a desire to include some debugging information when a rule crashes. This pull request uses the `debug` module to log the currently linted file, parser options, parser path, and shared plugin settings object, and could be augmented to include information on the node in question.

**TSC Question:** Should we accept this enhancement?

**Comment by wmertens** [comment]
Would it be an option to not re-throw the error, instead letting the rules that did not crash still do their thing?

**Comment by platinumazure** [comment]
> Would it be an option to not re-throw the error, instead letting the rules that did not crash still do their thing?

As long as we can figure out a way to report the error (including stack trace, etc.) while still reporting other role messages, I'm certainly not opposed, but I think we would need a concrete proposal. I could also see a CLI option to basically turn thrown errors into brief rule messages, trading debug information for the ability to keep running other rules.

I'm also okay with implementing this PR as a start and revisiting afterwards, as well.

**Comment by kaicataldo** [comment]
This issue was accepted at today's TSC meeting.

**Comment by pfhayes** [comment]
@platinumazure i added a test. I wasn't sure what the best practices around a more involved integration test would be, as I didn't see any similar tests in the repo. I'm happy to add further tests if any guidance is available

**Comment by not-an-aardvark** [comment]
Thanks for the PR! Was there a reason you added this logic to `traverser` rather than adding it to `Linter` directly? It seems like that could avoid the need to pass around `sharedTraversalContext`. (From a code cleanliness perspective, I'm a bit averse to passing around `sharedTraversalContext` because it's an implementation detail of `Linter`, and doesn't have all of the properties that one might expect from dealing with `context` objects in rules.)

To answer your question, there are a few integration tests in [`tests/bin/eslint.js`](https://github.com/eslint/eslint/blob/master/tests/bin/eslint.js), although in this case I'm not sure whether that would be any better because our debugging messages aren't considered to be a public API. I think the way you did it is generally fine.

**Comment by pfhayes** [comment]
I've moved the relevant code into the `Linter` class

File: lib/linter.js [hunk 1]
```diff
@@ -1000,16 +1001,27 @@ module.exports = class Linter {
...
+                    eventGenerator.enterNode(node);
+                },
+                leave(node) {
+                    eventGenerator.leaveNode(node);
+                },
+                visitorKeys: sourceCode.visitorKeys
+            });
+        } catch (err) {
+            this._logger("An error occurred while traversing");
+            if (this._context) {
```
Comment (not-an-aardvark) [inline lib/linter.js]: Was this a typo? I don't think `this._context` exists on `Linter`.

The filename, parser options, etc. should be in scope here already, so it's probably not necessary to pull them from `context`.

**Comment by pfhayes** [comment]
Good catch - I just updated them

File: lib/linter.js (hunk 1 above)
Comment (pfhayes) [reply lib/linter.js]: fixed, cc @not-an-aardvark

**Comment by not-an-aardvark** [comment]
@platinumazure I'm a bit worried about the need to expose the logger functionality to users of `Linter` for testing purposes -- we've had issues in the past with people monkeypatching `Linter` to get unintended functionality, which prevented us from changing things afterwards. I'm wondering if a test is actually necessary here, given that we don't have tests for any of our other debugging lines. (I'm also thinking this should have been labeled as `chore` and not needed TSC approval, since the TSC generally doesn't need to approve `debug` lines when they're added as part of other features.)

**Comment by platinumazure** [comment]
Okay, I can buy that we don't want to expose things on `Linter` and that we don't need a test.

I think whether this needed TSC discussion is debatable. I don't see this as a chore. To me, this was an enhancement that deserved some discussion on whether the benefit was worth the maintenance overhead, like any other enhancement to core. I'm glad we decided it was.

**Comment by not-an-aardvark** [comment]
I suppose my thinking is that this isn't a public API, so if we decided later that it wasn't worth the maintenance overhead, we could just remove it.

**Comment by platinumazure** [comment]
Hi @pfhayes, sorry we lost track of this issue.

Would you mind rebasing to fix the merge conflict currently in this PR? Then hopefully @not-an-aardvark can take a look at your changes since his last review and then hopefully we can merge this in. Thanks!

**Comment by pfhayes** [comment]
@platinumazure @not-an-aardvark updated

File: lib/linter.js [hunk 2]
```diff
@@ -908,6 +908,7 @@ module.exports = class Linter {
 
         this.rules = new Rules();
         this.environments = new Environments();
+        this._logger = debug;
```
Comment (kaicataldo) [inline lib/linter.js]: Just for my own understanding - is the reasoning for creating this property to make it testable?

**Comment by platinumazure** [comment]
That was the original intent, I think. I had requested that we write a test
case but I've been persuaded since then that it's not critical. I would
favor removing the property at this point, personally.

Not to speak for the PR author- just wanted to provide my perspective since
it was mostly at my request originally.

**Comment by kaicataldo** [comment]
Agreed - my preference would be to use debug directly here and not add another property.

**Comment by ljharb** [comment]
@platinumazure i'd recommend gitignoring all three common lockfile formats, to prevent them from being accidentally included in PRs and breaking CI builds.

**Comment by platinumazure** [comment]
@ljharb Good call. yarn.lock, package-lock.json, what's the third? npm-shrinkwrap.json?

**Comment by not-an-aardvark** [comment]
Would ignoring those files cause an issue if someone has a local version which is out of date? I had been thinking we should just add a test that fails if those files exist.

**Comment by pfhayes** [comment]
@not-an-aardvark @platinumazure thanks, I've made the requested updates

**Comment by platinumazure** [comment]
LGTM, thanks!

@not-an-aardvark @kaicataldo Does this look good on your end?

**Comment by not-an-aardvark** [comment]
Thanks for contributing!

**Pull request merged**